### DISCUSSION TITLE
Get Cmake conan dependencies from conan?

### DISCUSSION THREAD

**Discussion by avetiso** [mesonbuild/meson#10220] [Q&A]
I'm using Conan's meson integration, and I'm at the point where I need to get the dependencies into Meson.

Conan generates conanbuildinfo.txt which contains this important section:

```
[includedirs]
/Users/dario.avetisov/.conan/data/active_record/1.53.3/mx/stable/package/0025a6d31b54a7fa6690eaf4b7f07dcac703754d/include
/Users/dario.avetisov/.conan/data/active_remote/2.36.2/mx/stable/package/3c8e841d55716239a9911665d1ffd7bc43cac6de/include
/Users/dario.avetisov/.conan/data/checkmate/1.36.6/mx/stable/package/2b2a4eb93a8d80ad9fdf6c2d032a1cca001d7b2c/include
/Users/dario.avetisov/.conan/data/locale/1.39.2/mx/stable/package/a727b834c705c7da7d207a8ab5c47ea11badfedc/include
/Users/dario.avetisov/.conan/data/queue/1.32.2/mx/stable/package/98501773423965b658c32de4719df520dd175e60/include
/Users/dario.avetisov/.conan/data/common/1.34.2/mx/stable/package/41276176b6768f126e9a0f0185572e212f974cbf/include
/Users/dario.avetisov/.conan/data/sqlite3/3.26.0-6/mx/stable/package/69396a1e403c24f601452a2d014779f30099e4ab/include
/Users/dario.avetisov/.conan/data/curl/7.79.1-0/mx/stable/package/b46591a7948af4a09fe30c9af4096239197051f6/include
/Users/dario.avetisov/.conan/data/opencv/3.4.11-5/mx/stable/package/7a6cc83fade841338b570c6a832371feb941349d/include
/Users/dario.avetisov/.conan/data/tiny_dnn/2017.8.25-4/mx/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include
/Users/dario.avetisov/.conan/data/base64/1.0.0-5/mx/stable/package/69396a1e403c24f601452a2d014779f30099e4ab/include
/Users/dario.avetisov/.conan/data/openssl/1.1.1k-1/mx/stable/package/69396a1e403c24f601452a2d014779f30099e4ab/include
/Users/dario.avetisov/.conan/data/zlib/1.2.11-8/mx/stable/package/65b5071e1e8fefd56dd859ce0171993880d17623/include
/Users/dario.avetisov/.conan/data/cereal/1.2.2-2/mx/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include

[libdirs]
/Users/dario.avetisov/.conan/data/active_record/1.53.3/mx/stable/package/0025a6d31b54a7fa6690eaf4b7f07dcac703754d/lib
/Users/dario.avetisov/.conan/data/active_remote/2.36.2/mx/stable/package/3c8e841d55716239a9911665d1ffd7bc43cac6de/lib
/Users/dario.avetisov/.conan/data/checkmate/1.36.6/mx/stable/package/2b2a4eb93a8d80ad9fdf6c2d032a1cca001d7b2c/lib
/Users/dario.avetisov/.conan/data/locale/1.39.2/mx/stable/package/a727b834c705c7da7d207a8ab5c47ea11badfedc/lib
/Users/dario.avetisov/.conan/data/queue/1.32.2/mx/stable/package/98501773423965b658c32de4719df520dd175e60/lib
/Users/dario.avetisov/.conan/data/common/1.34.2/mx/stable/package/41276176b6768f126e9a0f0185572e212f974cbf/lib
/Users/dario.avetisov/.conan/data/sqlite3/3.26.0-6/mx/stable/package/69396a1e403c24f601452a2d014779f30099e4ab/lib
/Users/dario.avetisov/.conan/data/curl/7.79.1-0/mx/stable/package/b46591a7948af4a09fe30c9af4096239197051f6/lib
/Users/dario.avetisov/.conan/data/opencv/3.4.11-5/mx/stable/package/7a6cc83fade841338b570c6a832371feb941349d/lib
/Users/dario.avetisov/.conan/data/base64/1.0.0-5/mx/stable/package/69396a1e403c24f601452a2d014779f30099e4ab/lib
/Users/dario.avetisov/.conan/data/openssl/1.1.1k-1/mx/stable/package/69396a1e403c24f601452a2d014779f30099e4ab/lib
/Users/dario.avetisov/.conan/data/zlib/1.2.11-8/mx/stable/package/65b5071e1e8fefd56dd859ce0171993880d17623/lib
```

Which points to exactly where the includes and libs are for the versions of the package I am downloading from conan.

How do I bridge that to Meson? These are all Cmake packages.

**Comment by eli-schwartz** [comment]
At least a few of those definitely have pkg-config files available. curl, zlib, sqlite3, openssl...

So your conan environment should be publicizing a pkgconfig directory such that external projects can then find the conan built versions of those dependencies.

If there is no pkg-config integration, they may provide cmake find_package() compatible dependency lookups. Meson supports those too, using the same syntax as pkg-config.

If that still fails, you can use compiler.find_library, see the second note at https://mesonbuild.com/Reference-manual_returned_compiler.html for advice on extending the search path.

**Reply by avetiso** [reply]
We use Conan for everything, so I'm trying to find a pattern for finding and consuming these libs that works for all of our packages. Trying to find a way to consume those paths in the file I posted in such a way that I don't have to spell out the specific hashes in the meson.build, because the hashes will be different depending on the settings that the package was built with (OS, arch, compiler, etc.) Not sure if this is a meson problem to or a conan problem. Perhaps it makes sense for conan to generate something as part of their meson toolchain that meson can consume to find the libraries easily...

**Reply by eli-schwartz** [reply]
What would you do in order to get those Conan libraries to he detected by a Makefile that runs `gcc foo.c -l<libname> -o foo`?

How does Conan get the libraries onto the `-l` search path?

**Reply by avetiso** [reply]
In the context of CMake, conan generates CMake Files with the paths to the dependencies that just slots right into existing CMake build systems. I'm guessing that Meson needs a similar setup from Conan, aka Conan generates a file or calls meson with the flags that correspond exactly to the dependencies specified in the conan file.

**Reply by eli-schwartz** [reply]
What "CMake Files" with the paths to the dependencies does conan generate? Existing CMake build systems don't try to read any particular kind of file, they just do `find_package()` dependency lookups. Does conan generate a cmake toolchain (comparable to a meson `--native-file`)? Does conan generate its own `*-config.cmake` files in a central location that load these dependencies?

**Comment by mensinda** [comment]
For CMake deps specifically, Conan would have to set the [CMAKE_PREFIX_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_PREFIX_PATH.html) to the specific CMake deps. For Meson, this could be done via [cross and native files](https://mesonbuild.com/Machine-files.html#cmake-variables) or via `-Dcmake_prefix_path=` via the commandline.

**Reply by avetiso** [reply]
CMAKE_PREFIX_PATH would be difficult because I need to ensure that meson gets the exact hash that conan specifies, and that is part of the prefix, and the prefix would be different for every dependency as you can see from the example file I posted.

**Reply by eli-schwartz** [reply]
So you just need to set CMAKE_PREFIX_PATH to the combined list of prefixes that conan specifies? I don't see the problem.

**Reply by avetiso** [reply]
@eli-schwartz ah that's just my ignorance regarding cmake I think, I figured it just allowed one prefix, not a list.

In that case, I think I'll need to implement something for meson on the conan side. What do you think makes the most sense? The Meson toolchain for conan simply generates a crossfile with all the specific locations of the dependencies from the conanfile? What do you think is the most appropriate way for meson to consume such info? Cross file or command line arg?

**Reply by eli-schwartz** [reply]
I don't know.

If the idea is for conan to generate this info and then allow you to manually invoke your own builds while inside a conan environment / relying on conan deps, then a native/cross file is nice because it allows you to consolidate the relevant arguments into one toolchain file and just insert it into your build setup. Effectively aliases.

If the idea is for conan to expose this info to another conan package -- a conan package which builds your software using meson -- then conan presumably controls the `meson setup` command line and can just insert the command line arg(s) in the backend logic.
